lib/core/util.c

```diff
@@ -64,51 +64,159 @@ static void free_accept_data(struct st_h2o_accept_data_t *data)
+        h2o_memcached_get(async_resumption_context.memcached.ctx, data->ctx->libmemcached_receiver, session_id, memcached_resumption_on_get,
                           data, H2O_MEMCACHED_ENCODE_KEY | H2O_MEMCACHED_ENCODE_VALUE);
 }
 
-static void async_resumption_new(h2o_iovec_t session_id, h2o_iovec_t session_data)
+static void memcached_resumption_new(h2o_socket_t *sock, h2o_iovec_t session_id, h2o_iovec_t session_data)
 {
-    h2o_memcached_set(async_resumption_context.memc, session_id, session_data,
+    h2o_memcached_set(async_resumption_context.memcached.ctx, session_id, session_data,
                       (uint32_t)time(NULL) + async_resumption_context.expiration,
                       H2O_MEMCACHED_ENCODE_KEY | H2O_MEMCACHED_ENCODE_VALUE);
 }
 
-static void async_resumption_remove(h2o_iovec_t session_id)
+void h2o_accept_setup_memcached_ssl_resumption(h2o_memcached_context_t *memc, unsigned expiration)
 {
-    h2o_memcached_delete(async_resumption_context.memc, session_id, H2O_MEMCACHED_ENCODE_KEY);
+    async_resumption_context.memcached.ctx = memc;
+    async_resumption_context.expiration = expiration;
+    h2o_socket_ssl_async_resumption_init(memcached_resumption_get, memcached_resumption_new, NULL);
+}
+
+static void dispose_redis_connection(void *conn)
+{
+    h2o_redis_free((h2o_redis_conn_t *)conn);
+}
+
+h2o_redis_conn_t **redis_connection_for_context(h2o_context_t *ctx)
+{
+    static size_t key = SIZE_MAX;
+    return (h2o_redis_conn_t **)h2o_context_get_storage(ctx, &key, dispose_redis_connection);
+}
+
+static void on_redis_connect(void *data)
+{
+    fprintf(stderr, "connected to redis at %s:%" PRIu16 "\n", async_resumption_context.redis.host.base, async_resumption_context.redis.port);
+}
+
+static void on_redis_disconnect(void *data)
+{
+    fprintf(stderr, "disconnected from redis at %s:%" PRIu16 "\n", async_resumption_context.redis.host.base, async_resumption_context.redis.port);
+}
+
+static void on_redis_connection_failure(const char *errstr, void *data)
+{
+    fprintf(stderr, "redis connection failure: %s\n", errstr);
+
+    /* clear storage by setting NULL */
+    h2o_context_t *ctx = data;
+    h2o_redis_conn_t **conn = redis_connection_for_context(ctx);
+    h2o_redis_free(*conn);
+    *conn = NULL;
+}
+
+static h2o_redis_conn_t *get_redis_connection(h2o_context_t *ctx)
+{
+    h2o_redis_conn_t **conn = redis_connection_for_context(ctx);
+    if (*conn == NULL) {
+        h2o_redis_callbacks_t callbacks = (h2o_redis_callbacks_t){
+            on_redis_connect,
+            on_redis_disconnect,
+            on_redis_connection_failure,
+        };
+        *conn = h2o_redis_connect(ctx->loop, async_resumption_context.redis.host.base, async_resumption_context.redis.port, callbacks, ctx);
+    }
+
+    return *conn;
 }
 
-void h2o_accept_setup_async_ssl_resumption(h2o_memcached_context_t *memc, unsigned expiration)
+static h2o_iovec_t base64_encode(h2o_iovec_t str, h2o_iovec_t prefix)
```

Comment:
Please rename this function to a more appropriate name.

The name doesn't reflect what it is doing: _concatenate a prefix_ with a base64-encoded string.

One possible candidate would be: build_key.